File: Dynamic_Programming/1395. Count Number of Teams.cpp
/*
1395. Count Number of Teams
*/

// Array: 前後綴分解 類似瞻前顧後 TC:O(n^2) SC:O(1)
class Solution {
public:
   int numTeams(vector<int>& rating) {
        int res = 0;
        for (auto i = 1; i < rating.size() - 1; ++i) {
            int less[2] = {}, greater[2] = {};
            for (auto j = 0; j < rating.size(); ++j) {
                if (rating[i] < rating[j])
                    ++less[j > i];
                if (rating[i] > rating[j])
                    ++greater[j > i];
            }
            res += less[0] * greater[1] + greater[0] * less[1];
        }
        return res;
    }
};

// DP TC:O(n^2) SC:O(n)
class Solution {
public:
    int numTeams(vector<int>& rating) {
        int n = rating.size();
        vector dp(n+1, vector<vector<int>>(4, vector<int>(2, 0)));
        int ans = 0;
        for(int i = 0; i < n; i++) {
            dp[i+1][1][1] = 1;
            dp[i+1][1][0] = 1;
            for(int j = 0; j < i; j++) {
                if(rating[i] > rating[j]) {
                    dp[i+1][3][0] += dp[j+1][2][0];
                    dp[i+1][2][0] += dp[j+1][1][0];
                }
                else {
                    dp[i+1][3][1] += dp[j+1][2][1];
                    dp[i+1][2][1] += dp[j+1][1][1];
                }
            }
            ans += dp[i+1][3][0];
            ans += dp[i+1][3][1];
        }
        return ans;
    }
};

File: Data_Structure/1395. Count Number of Teams.py
# TC:O(nlgn) SC:O(n)
class FenwickTree:
    def __init__(self, n: int):
        self.tree = [0] * (n + 1) 
    def update(self, i: int, val: int) -> None:
        while i < len(self.tree):
            self.tree[i] += val
            i += i & -i

    def pre(self, i: int) -> int:
        res = 0
        while i > 0:
            res += self.tree[i]
            i &= i - 1
        return res
    def query(self, l: int, r: int) -> int:
        if r < l:
            return 0
        return self.pre(r) - self.pre(l - 1)
class Solution:
    def numTeams(self, rating: List[int]) -> int:
        '''
        也就是要你找遞增三元組 or 遞減三元組有多少個 
        1. python SortedList
        2. BIT
        '''
        n = len(rating)
        # 離散化
        nums = rating.copy()
        nums.sort()
        mp = {x: i + 1 for i, x in enumerate(nums)}  # 離散化 + BIT index 1-based

        bit = FenwickTree(n)
        ans = 0
        for i in range(n):
            idx = mp[rating[i]]
            # 查找有多少數字比我小? bit.pre(idx + 1) 然後判斷一下原本有多少數字比他大 n - bisect_right(nums, rating[i])
            y = bisect_left(nums, rating[i])
            ls = bit.pre(idx - 1) # 代表有 i - ls 的數字比我大 @ 左邊 所以右邊就有 (n - bisect_right(nums, rating[i])) - (i - ls) 比我大
            ll = i - ls
            rl = n - y - 1 - (i - ls)
            rs = y - ls# 右邊比我小的 目前有多少數字比我小 ls 總共有多少數字比我小 bisect_right(nums, rating[i]) -> bisect_right(nums, rating[i]) - ls 就是右邊的量
            ans += ls * rl + ll * rs
            bit.update(idx, 1)
        return ans

        '''
        SortedList
        sl_left = SortedList([rating[0]])
        sl_right = SortedList(rating[i] for i in range(1, n))
        ans = 0
        for i in range(1, n - 1):
            sl_right.remove(rating[i])
            ls = sl_left.bisect_left(rating[i]) 
            ll = len(sl_left) - ls
            rs = sl_right.bisect_left(rating[i])
            rl = len(sl_right) - rs
            ans += ls * rl + ll * rs
            sl_left.add(rating[i])
        return ans
        '''
